tools/unitctl/unitctl/src/cmd/instances.rs

```diff
@@ -1,16 +1,136 @@
 use crate::{OutputFormat, UnitctlError};
+use crate::unitctl_error::ControlSocketErrorKind;
+
+use std::path::PathBuf;
+use unit_client_rs::unitd_docker::deploy_new_container;
 use unit_client_rs::unitd_instance::UnitdInstance;
+use unit_client_rs::control_socket_address::ControlSocket;
+
+pub(crate) async fn cmd(args: InstanceArgs) -> Result<(), UnitctlError> {
+    if let Some(cmd) = args.command {
+        match cmd {
+            InstanceCommands::New {
+                ref socket,
+                ref application,
+                ref image,
+            } => {
+                // validation for application dir
+                if !PathBuf::from(application).is_dir() {
+                    eprintln!("application path must be a directory");
+                    Err(UnitctlError::NoFilesImported)
+                } else if !PathBuf::from(application).as_path().exists() {
+                    eprintln!("application path must exist");
+                    Err(UnitctlError::NoFilesImported)
+
+                } else {
+                    let addr = ControlSocket::parse_address(socket);
+                    if let Err(e) = addr {
+                        return Err(UnitctlError::UnitClientError{source: e});
+                    }
+
+                    // validate we arent processing an abstract socket
+                    if let ControlSocket::UnixLocalAbstractSocket(_) = addr.as_ref().unwrap() {
+                        return Err(UnitctlError::ControlSocketError{
+                            kind: ControlSocketErrorKind::General,
+                            message: "cannot pass abstract socket to docker container".to_string(),
+                        })
+                    }
+
+                    // warn user of OSX docker limitations
+                    if let ControlSocket::UnixLocalSocket(ref sock_path) = addr.as_ref().unwrap() {
+                        if cfg!(target_os = "macos") {
+                            return Err(UnitctlError::ControlSocketError{
+                                kind: ControlSocketErrorKind::General,
+                                message: format!("Docker on OSX will break unix sockets mounted {} {}",
+                                                 "in containers, see the following link for more information",
+                                                 "https://github.com/docker/for-mac/issues/483"),
+                            })
+                        }
+
+                        if !sock_path.is_dir() {
+                          return Err(UnitctlError::ControlSocketError{
+                                kind: ControlSocketErrorKind::General,
+                                message: format!("user must specify a directory of UNIX socket directory"),
+                            })
+                        }
+                    }
+
+                    // validate a TCP URI
+                    if let ControlSocket::TcpSocket(uri) = addr.as_ref().unwrap() {
```

Comment:
Would this be a candidate for `expect()` instead of `unwrap()` here? I only offer that option because it can simplify debugging by outputting a message that can be more easily searched in source code.

Reply:
again, the possibility of addr being Err(...) is ruled out on line 28.